### PR TITLE
[dynamo] Replace InstructionTranslator.checkpoint with speculate/restart

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112962
* #112961
* #111726
* #111725
* #111415
* #111717
* #111306
* #112921
* __->__ #112902

In my work on making guards installed eagerly (look up the stack), I found that our checkpoint/restore mechanism is very broken.  There is lots of state (especially in shape_env) which we don't checkpoint and restore properly.  We also have lots of mutable state on variable trackers already which is not checkpointed/restored.  (See other PRs in this stack for some spot fixes.)

Since we wanted to get rid of this anyway for making VariableTracker mutable, I figured I would just switch to restarting analysis.

For other usages of copy_graphstate/restore_graphstate:
1) Many usages were pointless and not needed, these are removed in PRs below this.
2) Some other usage (similar to this one) is removed in PRs above this.
3) The tricky one I am not handling is higher_order_ops, which uses checkpoint/restore a lot.    There might be some cases there where this speculate/restart trick won't work.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

Differential Revision: [D51040145](https://our.internmc.facebook.com/intern/diff/D51040145)

### CODE REVIEW

**Comment by voznesenskym** [comment]
> The tricky one I am not handling is higher_order_ops, which uses checkpoint/restore a lot. There might be some cases there where this speculate/restart trick won't work.

I can take a look at this one.

File: torch/_dynamo/symbolic_convert.py [hunk 1]
```diff
@@ -130,6 +130,45 @@
...
+    def fail_and_restart_analysis(self):
+        """
+        Start tracing of the current frame over again, and don't take this branch.
+        """
+        self.failed = True
+        raise exc.SpeculationRestartAnalysis()
+
+
+@dataclasses.dataclass
+class SpeculationLog:
```
Comment (voznesenskym) [inline torch/_dynamo/symbolic_convert.py:148]: nit quick doc on what this does

File: torch/_dynamo/exc.py [hunk 2]
```diff
@@ -41,6 +41,14 @@ class RestartAnalysis(TorchDynamoException):
     pass
 
 
+class SpeculationRestartAnalysis(RestartAnalysis):
+    pass
+
+
+class UnspecializeRestartAnalysis(RestartAnalysis):
```
Comment (voznesenskym) [inline torch/_dynamo/exc.py:48]: Is this handled anywhere specifically? Or are you just using the type for more contextual info?

Comment (jansel) [reply torch/_dynamo/exc.py:48]: It is just to give more context info.  Also it allows someone in the future to understand the reason for a restart.

**Review by voznesenskym — APPROVED** [review]

File: torch/_dynamo/exc.py (hunk 2 above)
Comment (jansel) [reply torch/_dynamo/exc.py:48]: Actually turns out I did need to use this to a fix a bug, so I guess good forsight on my part...

**Pull request closed without merging**